workflow-ui/backstack-android/src/androidTest/java/com/squareup/workflow1/ui/backstack/test/BackStackContainerTest.kt

```diff
@@ -0,0 +1,175 @@
+    lateinit var firstScreen: TestRendering
+
+    scenario.onActivity { activity ->
+      // Set some view state to be saved and restored.
+      activity.currentTestView.viewState = "hello world"
+      firstScreen = activity.backstack!!.top
+    }
+
+    // Navigate to another screen.
+    setBackstack(firstScreen, TestRendering("new screen"))
+    assertThat(viewForScreen("new screen").viewState).isEqualTo("")
+
+    // Destroy and recreate the activity.
+    scenario.recreate()
+
+    // Navigate back.
+    setBackstack(firstScreen)
+
+    // Check that the view state was actually restored.
+    assertThat(viewForScreen("initial").viewState).isEqualTo("hello world")
+  }
+
+  @Test fun initialization_order() {
+    val events = mutableListOf<String>()
+    fun log(name: String, event: String, view: ViewStateTestView) {
+      events += "$name $event viewState=${view.viewState}"
+    }
+
+    fun consumeEvents() = events.toList().also { events.clear() }
+
+    fun testRendering(name: String) = TestRendering(
+      name = name,
+      onViewCreated = { view -> log(name, "onViewCreated", view) },
+      onShowRendering = { view -> log(name, "onShowRendering", view) },
+      viewHooks = object : ViewHooks {
+        override fun onSaveInstanceState(view: ViewStateTestView) = log(name, "onSave", view)
+        override fun onRestoreInstanceState(view: ViewStateTestView) = log(name, "onRestore", view)
+        override fun onAttach(view: ViewStateTestView) = log(name, "onAttach", view)
+        override fun onDetach(view: ViewStateTestView) = log(name, "onDetach", view)
+      }
+    )
+
+    val firstRendering = testRendering("first")
+    val secondRendering = testRendering("second")
+
+    // Setup initial screen.
+    setBackstack(firstRendering)
+    waitForScreen(firstRendering.name)
+    scenario.onActivity {
+      it.currentTestView.viewState = "hello"
+    }
+    assertThat(consumeEvents()).containsExactly(
+      "first onViewCreated viewState=",
+      "first onShowRendering viewState=",
+      "first onShowRendering viewState=",
+      "first onAttach viewState="
+    )
+
+    // Navigate forward.
+    setBackstack(firstRendering, secondRendering)
+    waitForScreen(secondRendering.name)
+    assertThat(consumeEvents()).containsExactly(
+      "second onViewCreated viewState=",
+      "second onShowRendering viewState=",
+      "second onShowRendering viewState=",
+      "first onSave viewState=hello",
+      "first onDetach viewState=hello",
+      "second onAttach viewState=",
+      "first onAttach viewState=hello",
+      "first onDetach viewState=hello"
+    )
+
+    // Navigate back.
+    setBackstack(firstRendering)
+    waitForScreen(firstRendering.name)
+    assertThat(consumeEvents()).containsExactly(
+      "first onViewCreated viewState=",
+      "first onShowRendering viewState=",
+      "first onShowRendering viewState=",
+      "first onRestore viewState=hello",
```

Comment:
Note that the first `showRendering` happens before state is restored. @rjrjr is this intentional? Is it required? I'm guessing yes, since Activity will only emit onRestoreInstanceState after onStart, at which point all the views should have received their initial values, so we want the same ordering here. Makes sense that you'd only want to restore scroll position after the scrollable content had been initialized too.

Reply:
Very intentional, for the reasons you said. Otherwise the rendering always
stomps the restore. Think about config change in particular. The rendering
is usually something that has already been shown in the previous config,
and me have been typed over since then.

Reply:
And be assured tests will fail if you change it. There's pretty good coverage here.